src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.CoreRT.cs

```diff
@@ -331,9 +363,66 @@ public object GetOrRegisterObjectForComInstance(IntPtr externalComObject, Create
             if (flags.HasFlag(CreateObjectFlags.Aggregation))
                 throw new NotImplementedException();
 
-            object? wrapperMaybeLocal = wrapperMaybe;
-            retValue = null;
-            throw new NotImplementedException();
+            if (flags.HasFlag(CreateObjectFlags.Unwrap))
+            {
+                var comInterfaceDispatch = impl.TryGetComInterfaceDispatch(externalComObject);
+                if (comInterfaceDispatch != null)
+                {
+                    retValue = ComInterfaceDispatch.GetInstance<object>(comInterfaceDispatch);
+                    return true;
+                }
+            }
+
+            if (!flags.HasFlag(CreateObjectFlags.UniqueInstance))
+            {
+                using (LockHolder.Hold(impl._lock))
+                {
+                    if (impl._rcwCache.TryGetValue(externalComObject, out var handle))
+                    {
+                        if (handle.IsAllocated)
+                        {
+                            retValue = handle.Target;
+                            return false;
+                        }
+                    }
+                }
+            }
+
+            retValue = impl.CreateObject(externalComObject, flags);
+            if (retValue == null)
+            {
+                // If ComWrappers instance cannot create wrapper, we can do nothing here.
+                return false;
+            }
+
+            if (flags.HasFlag(CreateObjectFlags.UniqueInstance))
+            {
+                // No need to cache NativeObjectWrapper for unique instances. They are not cached.
+                return true;
+            }
+
+            using (LockHolder.Hold(impl._lock))
+            {
+                if (impl._rcwCache.TryGetValue(externalComObject, out var existingHandle))
+                {
+                    if (existingHandle.IsAllocated)
+                    {
+                        retValue = existingHandle.Target;
+                    }
+                }
+                else
+                {
+                    GCHandle proxyHandle = GCHandle.Alloc(retValue, GCHandleType.Weak);
```

Comment:
For symetry, it would be nice to allocated the GCHandle and add it to the table in NativeObjectWrapper constructor - since the opposite is done in the NativeObjectWrapper finalizer.